### DISCUSSION TITLE
Serializing rpc call bytes to send through websocket and reply on the other side

### DISCUSSION THREAD

**Discussion by germandiagogomez** [capnproto/capnproto#1887] [Q&A] [answered]
Hello everyone,

I am integrating capnproto in a client program that uses Wasm. Since it uses Wasm and will run in the browser, I need to serialize rpc calls through a websocket, arbitrary ports are not allowed. I already have the websocket infra working on the server side, but I am struggling to serialize calls from the client. I have something like this:

Client side: 

```
capnp::TwoPartyClient client(*connectionStream);
    auto accessClient =
        client.bootstrap().castAs<Guinyote::Capnp::OnlineRoomsAccess>();
    auto authRequest = accessClient.authenticateRequest();
    auto userInfo = authRequest.initUserInfo();
    userInfo.setNickName("SomeUser");

    // I want to flatten my request to send the bytes through a client websocket from client side
   // to forward them from the server websocket (on the remote side)
    capnp::messageToFlatArray(????);
```

Basically the flow I currently have is:

    1. Use my `Capnp::MyInterface::Client` and construct a request.
    2. Instead of sending the request directly to a server, flatten it to bytes
    3. Send the request through the local client websocket to the server
    4. on the server, a call to `wsMessageStream.receiveMessage()` is used to recover the message from the websocket.
    5.  the rpc call is forwarded to the real server through a socket (eventually could be a pipe)
    6. once the server reply is ready, it will be sent all the way back in reverse order, so I will need to serialize the reply.

The server reply could be a capability also, which needs to be sent back to the client by inserting it in the server websocket and send to the client websocket, etc.

I am a bit lost as to what to do exactly to make all this work since I am a bit lost on the serialization of rpcs and interfaces. 

Any hint or better strategy if this is not the best way  is welcome as well.

Probably this can be of good use, but not sure how to integrate it:

   - `AsyncCapabilityStream`
   - `CapabilityStreamConnectionReceiver`

Thanks for your time.

**Comment by kentonv** [comment]
> 2. Instead of sending the request directly to a server, flatten it to bytes

I'm afraid Cap'n Proto's RPC protocol is more complicated than this. A single RPC call does not necessarily map to a single message sent on the wire, and the system just isn't designed for you to be able to extract the messages at certain points in time like this. Instead, you must give the RPC system an implementation of `capnp::MessageStream` (or `kj::AsyncIoStream`) that lets it read and write messages as needed.

But doing this in Wasm might be tricky. KJ currently provides async event loop implementations for epoll (Linux), kqueue (macOS/*BSD), poll (other Unix), and Windows APIs. But it Wasm, you might need something different. So this might be a bit of a project to build out.

The other route you can take is to not use Cap'n Proto RPC, and instead focus just on the serialization format. Then you can design your protocol to send one message at a time at the times you expect, and control how that is sent.

Sorry, I realize this is probably not what you wanted to hear here.

**Comment by germandiagogomez** [comment]
> Instead, you must give the RPC system an implementation of capnp::MessageStream (or kj::AsyncIoStream) that lets it read and write messages as needed.

There is a websocketmessagestream class in websocket-rpc.h. Would that work at least in native desktop to stream rpc? And from there I can try to figure out as a first step.

**Reply by kentonv** [reply]
Yes, that'll work well in an application using kj-http to implement the WebSocket protocol. So you could use that on the server.

**Reply by germandiagogomez** [reply]
I am still struggling a bit on how to do the wiring to make everything work.

For an easier to follow discussion, I will refer to the diagram below:

![wsserver](https://github.com/capnproto/capnproto/assets/13551083/365af868-a02b-4b5e-8976-db583c0f6d35)


### What I am currently trying to do and questions

1. I connect my `CapnpClient` interface for `Client Websocket`, I build a request with the Capnproto API and after send() the `ClientWebsocket` will forward the request to `Server Websocket`.

Question: In the original connection (client connecting directly to server, via `TwoPartyClient` I could bootstrap the interface I need). If I connect to the Websocket proxy, will I be able to bootstrap the interface like below? 

```
TwoPartyClient client(someStreamThatGoesToWebsocketAndNotToServerDirectly);
auto interface = client.bootstrap().castAs<CapnpServerInterface>();
```
Alternatively, can I do it dynamically in some way? The problem is that I am not connecting to the server directly anymore.

Now, with proxy,  I do something like this (roughly):

```
auto caps = provider->newCapabilityPipe();
TwoPartyClient capnpclienttowebsocketclient(caps.ends[0]);
// Try to bootstrap interface, did not try yet though
```

2. On the server side, in `Server Websocket` proxy, I am using `capnp::WebSocketMessageStream::readMessage` for every message that comes in. The result from readMessage is a `MessageReader`. I need to forwardRequest to `CapnpServer` so I need the `MessageReader` message contents into `CapnpServer` but I have no idea how to achieve that or even if it is possible. I just want forwarding, I do not care about anything else.

For the response trip, same thing: I do not know how to keep writing messages back through the flow until it reaches the original client.

Maybe I am doing a ton of things wrong... not sure how to proceed.

Originally I connected the client directly to the server, I bootstrapped the interface, and everything worked. 

But now, with `Client WebSocket` in the middle my `CapnpClient` won't connect anymore to the server directly, so I am not sure how to bootstrap the interface I need (maybe dynamically in some way?). 

Also, once I `sendRequest` by building a rpc with capnproto API and doint `.send()` my request should be directed to `Client WebSocket`, and from there I need to forward that request to the other side of the websocket, but not sure how.

**Reply by germandiagogomez** [reply]
Basically what I want to achieve is as much as possible tunneling through websocket from my original client to the original server. That's all there is to it. Easier said than done though.

**Reply by kentonv** [reply, marked as answer]
You shouldn't be calling `WebSocketMessageStream::readMessage()` directly. What you want is to give the `MessageStream` to `TwoPartyServer` so it can read the messages directly.

However, I notice that `TwoPartyServer::accept()` only has overloads for byte streams, not `MessageStream`. `TwoPartyVatNetwork`, though, supports `MessageStream`, and `TwoPartyServer` is just a thin wrapper around `TwoPartyVatNetwork`, so this appears to just be a small missing feature... we need to add `TwoPartyServer::accept(MessageStream&)`.

Similarly, on the client side, `TwoPartyClient`'s constructor currently only accepts byte streams, but we should add an overload that accepts `MessageStream&`. But again, on the client side, in Wasm, you can't use KJ HTTP. You'd need to implement `MessageStream` yourself in a way that calls the browser's WebSocket API under the hood.

**Reply by germandiagogomez** [reply]
Hello. I switched to finishing the original clients since this seems to be a bit of work and I have no bandwidth to port part of capnproto to the web. My priority is finishing up the work first, given there are some challenges ahead.

When I release the desktop versions, I will consider web clients or mobile for my networked game. Until then, I consider the answer as accepted since I would say I have enough information to proceed in case I need it.

It seems the difficult or more problematic part will be, as you mentioned again in the reply, the web client loop inner workings. Also, thanks for clearing up how to proceed for the message stream and the fact that `readMessage` should never be called directly. That part was confusing me.

Thanks for your help! Very appreciated.
